src/pow.rs

```diff
@@ -0,0 +1,1314 @@
+    fn index(&self, index: I) -> &Self::Output { &self.0[index] }
+}
+
+crate::internal_macros::display_from_debug!(Uint256);
+
+impl Encodable for Uint256 {
+    fn consensus_encode<W: crate::io::Write + ?Sized>(
+        &self,
+        w: &mut W,
+    ) -> Result<usize, crate::io::Error> {
+        let &Uint256(ref data) = self;
+        let mut len = 0;
+        for word in data.iter() {
+            len += word.consensus_encode(w)?;
+        }
+        Ok(len)
+    }
+}
+
+impl Decodable for Uint256 {
+    fn consensus_decode<R: crate::io::Read + ?Sized>(
+        r: &mut R,
+    ) -> Result<Uint256, crate::consensus::encode::Error> {
+        let mut ret: [u64; 4] = [0; 4];
+        for word in &mut ret {
+            *word = Decodable::consensus_decode(r)?;
+        }
+        Ok(Uint256(ret))
+    }
+}
+
+#[cfg(feature = "serde")]
+#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
+impl crate::serde::Serialize for Uint256 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: crate::serde::Serializer,
+    {
+        use crate::hashes::hex::ToHex;
+        let bytes = self.to_be_bytes();
+        if serializer.is_human_readable() {
+            serializer.serialize_str(&bytes.to_hex())
+        } else {
+            serializer.serialize_bytes(&bytes)
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
+#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
+impl<'de> crate::serde::Deserialize<'de> for Uint256 {
+    fn deserialize<D: crate::serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        use crate::hashes::hex::FromHex;
+        use crate::serde::de;
+        struct Visitor;
+        impl<'de> de::Visitor<'de> for Visitor {
+            type Value = Uint256;
+
+            fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                write!(f, "32 bytes or a hex string with 64 characters")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                let bytes = crate::prelude::Vec::from_hex(s)
+                    .map_err(|_| de::Error::invalid_value(de::Unexpected::Str(s), &self))?;
+                Uint256::from_be_slice(&bytes)
+                    .map_err(|_| de::Error::invalid_length(bytes.len() * 2, &self))
+            }
+
+            fn visit_bytes<E>(self, bytes: &[u8]) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Uint256::from_be_slice(bytes)
+                    .map_err(|_| de::Error::invalid_length(bytes.len(), &self))
+            }
+        }
+
+        if deserializer.is_human_readable() {
+            deserializer.deserialize_str(Visitor)
+        } else {
+            deserializer.deserialize_bytes(Visitor)
+        }
+    }
+}
+
+/// Invalid slice length.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Hash)]
+pub struct ParseLengthError {
```

Comment:
If we're going to keep `from_*_slice` methods this should be called `FromSliceError` since the error has nothing to do with parsing text.

Reply:
oh I didn't even notice, during refactoring to use `TryFrom` and remove `_slice` methods this error type is no longer used.